### PR TITLE
[AIRFLOW-4048] HttpSensor provide-context to response_check

### PR DESCRIPTION
**Pull request by raphaelauv**
Make sure you have checked _all_ steps below.

### Jira

- [x] https://issues.apache.org/jira/browse/AIRFLOW-4048

### Description

- The httpSensor do not make any use of the context inside the poke function , also sensors frequently  need execution context to evaluate the response inside the response_check function.

By example a HttpSensor looking for availability of data thank to an API , is going to need the context["execution_date"] to compare the date of data availability.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [ ] Passes `flake8`

### CODE REVIEW

File: tests/sensors/test_http_sensor.py
Comment (Fokko) [inline tests/sensors/test_http_sensor.py]: I think it would be easy to make this test a bit more meaningful. We should check if certain parameters inside of the response are being set.

**Review by Fokko — COMMENTED** [review]
I like the PR, and I can see that this is useful. I think we should improve a bit on the test, since we're not really asserting anything.

File: tests/sensors/test_http_sensor.py
Comment (raphaelauv) [reply tests/sensors/test_http_sensor.py]: Yes , but in the tests the context is empty , the mock do not provide any context...

Comment (Fokko) [reply tests/sensors/test_http_sensor.py]: If we make it part of a dag, we should be able to get the execution date etc.

**Comment by raphaelauv** [comment]
Hello , I have updated the test , but  I have a weird error with the travis , ( maybe a race ... ) can we re-run the failing travis (  I do not have the right ) @Fokko ?

many thanks

**Comment by zhongjiajie** [comment]
@raphaelauv You could use `git push -f` re-submit commit and the failed CI will restart.

**Comment by zhongjiajie** [comment]
@raphaelauv  BTW, we are already remove `py2` in our CI, and you failed test is in `py2.7` part. Maybe we should check CI log after you rebase on master.

**Comment by zhongjiajie** [comment]
Can't found any specific error in CI log, maybe we should restart it. @raphaelauv

**Comment by OmerJog** [comment]
@zhongjiajie CI is green

**Comment by raphaelauv** [comment]
@Fokko Could you review the PR , so it could be merge , thank you again

File: tests/sensors/test_http_sensor.py
Comment (raphaelauv) [reply tests/sensors/test_http_sensor.py]: done

File: tests/sensors/test_http_sensor.py
Comment (Fokko) [inline tests/sensors/test_http_sensor.py]:
```suggestion
                    if context["execution_date"] == DEFAULT_DATE:
```

**Comment by Fokko** [comment]
Thanks @raphaelauv the test is much better now, I have one more suggestion.

File: airflow/sensors/http_sensor.py
Comment (Fokko) [inline airflow/sensors/http_sensor.py]:
```suggestion
                return self.response_check(response, context)
```
Would do the trick as well, right? This will break backward compatibility because if the context is undefined, this won't work.

I think we can make this nicer:
```python
if self.response_check:
  if self.provide_context:
    return self.response_check(response, context)
  else:
    return self.response_check(response)
```
WDYT?

**Review by Fokko — COMMENTED** [review]
Thanks for the quick response @raphaelauv I have one more comment that I just missed, would like to pick your brain on it.

File: airflow/sensors/http_sensor.py
Comment (raphaelauv) [reply airflow/sensors/http_sensor.py]: Yes we can do this

File: airflow/sensors/http_sensor.py
Comment (ashb) [inline airflow/sensors/http_sensor.py]: This behaviour doesn't match the description - (keyword-args) for the description to be true this line should be

```
                     return self.response_check(response, **context)
```

This diff here is probably the right thing as it resembles what the provide_context arg does for the PythonOperator.

An example in the docs would help clarify this:

```rst
    Executes a HTTP GET statement and returns False on failure caused by
    404 Not Found or `response_check` returning False.

    HTTP Error codes other than 404 (like 403) or Connection Refused Error
    would fail the sensor itself directly (no more poking).


    The response check can access the template context by passing ``provide_context=True`` to the operator::

        def response_check(response, **context):
            # Can look at context['ti'] etc.
            return True

        HttpSensor(task_id='my_http_sensor', ..., provide_context=True, response_check=response_check)


    :param http_conn_id: The connection to run the sensor against
    :type http_conn_id: str
    :param method: The HTTP request method to use
    :type method: str
    :param endpoint: The relative part of the full url
    :type endpoint: str
    :param request_params: The parameters to be added to the GET url
    :type request_params: a dictionary of string key/value pairs
    :param headers: The HTTP headers to be added to the GET request
    :type headers: a dictionary of string key/value pairs
    :param provide_context: if set to true, Airflow will pass a set of
        keyword arguments that can be used in your function. This set of
        kwargs correspond exactly to what you can use in your jinja
        templates.
    :type provide_context: bool
    :param response_check: A check against the 'requests' response object.
        Returns True for 'pass' and False otherwise.
    :type response_check: A lambda or defined function.
    :param extra_options: Extra options for the 'requests' library, see the
        'requests' documentation (options to modify timeout, ssl, etc.)
    :type extra_options: A dictionary of options, where key is string and value
        depends on the option that's being modified.
```

Probably want to format the example a bit better (line wrap etc.)

Comment (raphaelauv) [reply airflow/sensors/http_sensor.py]: This was my first proposition but @Fokko proposed another one ( you can check )

Comment (ashb) [reply airflow/sensors/http_sensor.py]: That was what you had before:

```python
response_check_kwargs["context"] = context
# ...
response_check(response, **response_check_kwargs)
```
which would still have the check fn having a signature of `(response, context)`. I am proposing making it have a signature of `(response, **context)`.

File: airflow/sensors/http_sensor.py
Comment (ashb) [inline airflow/sensors/http_sensor.py]:
```suggestion
    The response check can access the template context by passing ``provide_context=True`` to the operator::
```

The double `::` makes the next block a code-block

**Comment by raphaelauv** [comment]
All remarks applied and base branch switch to v1-10-test

**Comment by ashb** [comment]
@raphaelauv It needs to stay targeting the master branch please. Once it is merged then I or another release manager will cherry it on to the release branch.

**Review by ashb — APPROVED** [review]
Code looks good now.

**Comment by Fokko** [comment]
I see a lot of unrelated changes, I think a rebase went wrong.

**Comment by ashb** [comment]
@raphaelauv I updated your the branch and rebased it against master. If this passes tests I will merge

**Review by ashb — APPROVED** [review]

**Pull request merged**
